internal/clients/http/azure/create_vm.go

```diff
@@ -38,68 +37,113 @@ const (
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "BeginCreateVM")
 	defer span.End()
 
 	logger := logger(ctx)
-	logger.Debug().Msg("Creating Azure VM instance")
+	logger.Debug().Msg("Creating Azure VM instance without waiting")
 
-	imageID := c.getImageId(ctx, resourceGroupName, imageName)
+	vmAzureParams, err := c.prepareVMresources(ctx, vmParams, vmName)
+	if err != nil {
+		return "", err
+	}
+
+	poller, err := c.beginCreateVM(ctx, vmParams.ResourceGroupName, vmName, vmAzureParams)
+	if err != nil {
+		span.SetStatus(codes.Error, "cannot create virtual machine")
+		logger.Error().Err(err).Msgf("cannot create virtual machine")
+		return "", err
+	}
+
+	resumeToken, err := poller.ResumeToken()
+	if err != nil {
+		span.SetStatus(codes.Error, "cannot generate resume token")
+		logger.Error().Err(err).Msgf("cannot generate the resume token")
+		return "", fmt.Errorf("cannot generate Azure resume token: %w", err)
+	}
+
+	return resumeToken, nil
+}
+
+func (c *client) WaitForVM(ctx context.Context, resumeToken string) (clients.AzureInstanceID, error) {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "WaitForVM")
+	defer span.End()
+
+	logger := logger(ctx)
+	logger.Debug().Msgf("Starting polling for Azure VM instance creation, using token: %s", resumeToken)
+
+	vmClient, err := c.newVirtualMachinesClient(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	pollerResponse, err := vmClient.BeginCreateOrUpdate(ctx, "", "", armcompute.VirtualMachine{}, &armcompute.VirtualMachinesClientBeginCreateOrUpdateOptions{
+		ResumeToken: resumeToken,
+	})
+	if err != nil {
+		span.SetStatus(codes.Error, "polling of virtual machine creation status failed to start")
+		return "", fmt.Errorf("polling of virtual machine creation status failed to start: %w", err)
+	}
+	resp, err := pollerResponse.PollUntilDone(ctx, nil)
+	if err != nil {
+		span.SetStatus(codes.Error, "failed to poll for create virtual machine status")
+		return "", fmt.Errorf("failed to poll for create virtual machine status: %w", err)
+	}
+
+	logger.Debug().Msgf("Done creating virtual machine id=%s", *resp.VirtualMachine.ID)
+
+	return clients.AzureInstanceID(*resp.VirtualMachine.ID), nil
+}
+
+func (c *client) prepareVMresources(ctx context.Context, vmParams clients.AzureInstanceParams, vmName string) (*armcompute.VirtualMachine, error) {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "prepareVMresources")
+	defer span.End()
 
-	virtualNetwork, err := c.createVirtualNetwork(ctx, location, resourceGroupName, vnetName)
+	logger := logger(ctx)
+	imageID := c.getImageId(ctx, vmParams.ResourceGroupName, vmParams.ImageName)
+
+	virtualNetwork, err := c.createVirtualNetwork(ctx, vmParams.Location, vmParams.ResourceGroupName, vnetName)
 	if err != nil {
 		span.SetStatus(codes.Error, "cannot create virtual network")
-		logger.Error().Err(err).Msg("cannot create virtual network")
+		logger.Error().Err(err).Msgf("cannot create virtual network")
```

Comment:
Why do you do that? :-) `Msg` is actually correct, I am not sure if `Msgf` would even work. Probably yes.

You have this at multiple places.

Reply:
no idea, bad rebase I'm guessing... let me fix that :)